File: CS/NonPersistentObjectsDemo.Module/BusinessObjects/PersistentObjects.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Xpo;

namespace NonPersistentObjectsDemo.Module.BusinessObjects {

    [DefaultClassOptions]
    [DevExpress.ExpressApp.DC.XafDefaultProperty(nameof(Order.Address))]
    public class Order : BaseObject {
        public Order(Session session) : base(session) { }

        private DateTime _Date;
        public DateTime Date {
            get { return _Date; }
            set { SetPropertyValue<DateTime>(nameof(Date), ref _Date, value); }
        }
        private string _Address;
        public string Address {
            get { return _Address; }
            set { SetPropertyValue<string>(nameof(Address), ref _Address, value); }
        }
        private decimal _Total;
        public decimal Total {
            get { return _Total; }
            set { SetPropertyValue<decimal>(nameof(Total), ref _Total, value); }
        }
        private OrderStatus _Status;
        public OrderStatus Status {
            get { return _Status; }
            set { SetPropertyValue<OrderStatus>(nameof(Status), ref _Status, value); }
        }
        [Aggregated]
        [Association]
        public XPCollection<OrderLine> Details {
            get { return GetCollection<OrderLine>(nameof(Details)); }
        }
    }

    public enum OrderStatus {
        Pending, Confirmed, Ready, Delivered, Canceled
    }

    [DevExpress.ExpressApp.DC.XafDefaultProperty(nameof(OrderLine.Product))]
    public class OrderLine : BaseObject {
        public OrderLine(Session session) : base(session) { }

        private Order _Order;
        [Association]
        public Order Order {
            get { return _Order; }
            set { SetPropertyValue<Order>(nameof(Order), ref _Order, value); }
        }
        private Product _Product;
        public Product Product {
            get { return _Product; }
            set { SetPropertyValue<Product>(nameof(Product), ref _Product, value); }
        }
        private int _Quantity;
        public int Quantity {
            get { return _Quantity; }
            set { SetPropertyValue<int>(nameof(Quantity), ref _Quantity, value); }
        }
    }

    [DefaultClassOptions]
    public class Product : BaseObject {
        public Product(Session session) : base(session) { }

        private string _Name;
        public string Name {
            get { return _Name; }
            set { SetPropertyValue<string>(nameof(Name), ref _Name, value); }
        }
        private decimal _Price;
        public decimal Price {
            get { return _Price; }
            set { SetPropertyValue<decimal>(nameof(Price), ref _Price, value); }
        }
    }

}

File: VB/NonPersistentObjectsDemo.Module/BusinessObjects/PersistentObjects.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Xpo

Namespace NonPersistentObjectsDemo.Module.BusinessObjects

	<DefaultClassOptions>
	<DevExpress.ExpressApp.DC.XafDefaultProperty(NameOf(Order.Address))>
	Public Class Order
		Inherits BaseObject

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _Date As DateTime
		Public Property [Date]() As DateTime
			Get
				Return _Date
			End Get
			Set(ByVal value As DateTime)
				SetPropertyValue(Of DateTime)(NameOf([Date]), _Date, value)
			End Set
		End Property
		Private _Address As String
		Public Property Address() As String
			Get
				Return _Address
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(Address), _Address, value)
			End Set
		End Property
		Private _Total As Decimal
		Public Property Total() As Decimal
			Get
				Return _Total
			End Get
			Set(ByVal value As Decimal)
				SetPropertyValue(Of Decimal)(NameOf(Total), _Total, value)
			End Set
		End Property
		Private _Status As OrderStatus
		Public Property Status() As OrderStatus
			Get
				Return _Status
			End Get
			Set(ByVal value As OrderStatus)
				SetPropertyValue(Of OrderStatus)(NameOf(Status), _Status, value)
			End Set
		End Property
		<Aggregated>
		<Association>
		Public ReadOnly Property Details() As XPCollection(Of OrderLine)
			Get
				Return GetCollection(Of OrderLine)(NameOf(Details))
			End Get
		End Property
	End Class

	Public Enum OrderStatus
		Pending
		Confirmed
		Ready
		Delivered
		Canceled
	End Enum

	<DevExpress.ExpressApp.DC.XafDefaultProperty(NameOf(OrderLine.Product))>
	Public Class OrderLine
		Inherits BaseObject

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _Order As Order
		<Association>
		Public Property Order() As Order
			Get
				Return _Order
			End Get
			Set(ByVal value As Order)
				SetPropertyValue(Of Order)(NameOf(Order), _Order, value)
			End Set
		End Property
		Private _Product As Product
		Public Property Product() As Product
			Get
				Return _Product
			End Get
			Set(ByVal value As Product)
				SetPropertyValue(Of Product)(NameOf(Product), _Product, value)
			End Set
		End Property
		Private _Quantity As Integer
		Public Property Quantity() As Integer
			Get
				Return _Quantity
			End Get
			Set(ByVal value As Integer)
				SetPropertyValue(Of Integer)(NameOf(Quantity), _Quantity, value)
			End Set
		End Property
	End Class

	<DefaultClassOptions>
	Public Class Product
		Inherits BaseObject

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _Name As String
		Public Property Name() As String
			Get
				Return _Name
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(Name), _Name, value)
			End Set
		End Property
		Private _Price As Decimal
		Public Property Price() As Decimal
			Get
				Return _Price
			End Get
			Set(ByVal value As Decimal)
				SetPropertyValue(Of Decimal)(NameOf(Price), _Price, value)
			End Set
		End Property
	End Class

End Namespace
